tests/knapsack.egg

```diff
@@ -0,0 +1,54 @@
+(datatype expr 
+    (Num i64)
+    (Add expr expr)
+    (Max expr expr))
+(rewrite (Add (Num a) (Num b)) (Num (+ a b)))
+(rewrite (Max (Num a) (Num b)) (Num (max a b)))
+
+; List of (weight, value) pairs
+(datatype objects
+    (Cons i64 i64 objects)
+    (Nil))
+
+; Given a capacity and a list of objects, finds the maximum value of a
+; collection of objects whose total weight does not exceed the capacity.
+(function Knap (i64 objects) expr)
+
+(rule ((= f (Knap capacity (Cons weight val rest))) (<= weight capacity))
+      ((set (Knap capacity (Cons weight val rest))
```

Comment:
maybe just use `f` here since you have bound `(Knap capacity (Cons weight val rest))` to `f`

Reply:
Resolving from offline conversation - reusing f would work for `union` but not for `set`